test/object-store/realm.cpp

```diff
@@ -854,44 +872,122 @@ TEST_CASE("SharedRealm: async_writes") {
+            auto col = table->get_column_key("value");
+            table->create_object().set(col, 45);
+            done = true;
+            realm->async_commit_transaction([&]() {});
+        });
         realm->async_begin_transaction([&]() {
-            REQUIRE(write_nr == 0);
-            ++write_nr;
             auto table = realm->read_group().get_table("class_object");
             auto col = table->get_column_key("value");
             table->create_object().set(col, 45);
             realm->async_commit_transaction([&]() {
-                REQUIRE(commit_nr == 0);
-                ++commit_nr;
                 done = true;
             });
         });
-        REQUIRE_THROWS_WITH(
-            realm->begin_transaction(),
-            Catch::Matchers::Contains("Can't begin transaction while an async transaction is ongoing"));
         util::EventLoop::main().run_until([&] {
             return done;
         });
-        REQUIRE(done);
-        realm->begin_transaction(); // Now is ok
+        realm->begin_transaction(); // Here syncing of first async tr has not completed
+        auto table = realm->read_group().get_table("class_object");
+        REQUIRE(table->size() == 1);
+        auto col = table->get_column_key("value");
+        table->create_object().set(col, 90);
+        realm->commit_transaction(); // Will re-initiate async writes
+
+        done = false;
+        util::EventLoop::main().run_until([&] {
+            return done;
+        });
+        REQUIRE(table->size() == 3);
+    }
+    SECTION("asynchronous transaction mixed with sync transaction that is cancelled") {
+        bool persisted = false;
+        realm->async_begin_transaction([&]() {
+            auto table = realm->read_group().get_table("class_object");
+            auto col = table->get_column_key("value");
+            table->create_object().set(col, 45);
+            done = true;
+            realm->async_commit_transaction([&]() {
+                persisted = true;
+            });
+        });
+        realm->async_begin_transaction([&]() {
+            auto table = realm->read_group().get_table("class_object");
+            auto col = table->get_column_key("value");
+            table->create_object().set(col, 45);
+            realm->async_commit_transaction([&]() {
+                persisted = true;
+            });
+        });
+        util::EventLoop::main().run_until([&] {
+            return done;
+        });
+        realm->begin_transaction();
+        CHECK(persisted);
+        persisted = false;
+        auto table = realm->read_group().get_table("class_object");
+        REQUIRE(table->size() == 1);
+        auto col = table->get_column_key("value");
+        table->create_object().set(col, 90);
         realm->cancel_transaction();
+
+        util::EventLoop::main().run_until([&] {
+            return persisted;
+        });
+        REQUIRE(table->size() == 2);
```

Comment:
Maybe add a stronger check as well to make sure the write was from the async transaction and not the cancelled one?